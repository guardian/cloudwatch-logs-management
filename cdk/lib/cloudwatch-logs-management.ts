import { CfnInclude } from "@aws-cdk/cloudformation-include";
import type { App } from "@aws-cdk/core";
import type {
  GuStackProps,
  GuStageParameter,
} from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";

export class CloudwatchLogsManagement extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    new CfnInclude(this, "Template", {
      templateFile: "../template.yaml",
      parameters: {
        Stage: this.getParam<GuStageParameter>("Stage"),
      },
    });
  }
}
