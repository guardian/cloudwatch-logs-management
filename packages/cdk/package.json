{
  "name": "cdk",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "synth": "cdk synth --path-metadata false --version-reporting false",
    "diff": "cdk diff --path-metadata false --version-reporting false"
  },
  "devDependencies": {
    "@guardian/cdk": "50.11.3",
    "aws-cdk": "2.90.0",
    "aws-cdk-lib": "2.90.0",
    "constructs": "10.2.69",
    "source-map-support": "^0.5.20",
    "ts-node": "^10.9.1"
  },
  "jest": {
    "testMatch": [
      "<rootDir>/lib/**/*.test.ts"
    ],
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "setupFilesAfterEnv": [
      "./jest.setup.js"
    ]
  }
}
