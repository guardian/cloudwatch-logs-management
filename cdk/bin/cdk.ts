#!/usr/bin/env node

import "source-map-support/register";
import { App } from "@aws-cdk/core";
import { CloudwatchLogsManagement } from "../lib/cloudwatch-logs-management";

const app = new App();
//  TODO: Add stack name
//   e.g. { stack: "SomeStack" }
new CloudwatchLogsManagement(app, "Template", {
  stack: "cloudwatch-logs-management",
  migratedFromCloudFormation: true,
});
