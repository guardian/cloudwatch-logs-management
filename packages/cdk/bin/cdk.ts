import 'source-map-support/register';
import { GuRoot } from "@guardian/cdk/lib/constructs/root";
import { CloudwatchLogsManagement } from '../lib/cloudwatch-logs-management';
import {CloudwatchLogsRetention} from "../lib/cloudwatch-logs-retention";
import {GuStackProps} from "@guardian/cdk/lib/constructs/core";

const app = new GuRoot();

export interface CloudwatchLogsManagementProps
	extends Omit<GuStackProps, 'stage' | 'env'> {
	retentionInDays?: number;
	logShippingPrefixes?: string[];
}

export const retentionOnlyStacks: CloudwatchLogsManagementProps[] = [
	{
		stack: 'membership',
		retentionInDays: 14,
	}
];

export const retentionAndTransferStacks: CloudwatchLogsManagementProps[] = [
	{ stack: 'print-production' },
	{ stack: 'deploy' },
	{ stack: 'flexible' },
	{ stack: 'workflow' },
	{ stack: 'media-service',
		logShippingPrefixes: [
			'/aws/lambda',
			'/aws/transfer'
		],
	},
	{ stack: 'content-api' },
	{ stack: 'cms-fronts' },
	{ stack: 'ophan' },
	{ stack: 'frontend' },
	{ stack: 'identity' },
	{ stack: 'interactives' },
	{ stack: 'mobile' },
	{ stack: 'security' },
	{ stack: 'personalisation' },
	{
		stack: 'targeting',
		logShippingPrefixes: [
			'/aws/lambda/diff-checker-PROD',
			'/aws/lambda/diff-publisher-PROD',
			'/aws/lambda/braze-exporter-trigger-PROD',
			'/aws/lambda/braze-exporter-callback-PROD',
		],
	},
	{
		stack: 'pfi',
		logShippingPrefixes: [
			'/aws/lambda/lurch',
			'/aws/rds/instance/lurch',
			'/aws/lambda/pfi-lurch',
			'fb-ad-library',
			'lurch',
			'/aws/lambda/transcription-service',
		],
	},
	{ stack: 'playground' },
	{ stack: 'ai' },
];

export const allRetentionStacks = retentionOnlyStacks.concat(retentionAndTransferStacks);

allRetentionStacks.forEach((stack) => { new CloudwatchLogsRetention(app, stack)});

retentionAndTransferStacks.forEach((stack) => { new CloudwatchLogsManagement(app, stack)});

