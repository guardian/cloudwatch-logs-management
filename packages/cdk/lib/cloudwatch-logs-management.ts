import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { GuS3Bucket } from '@guardian/cdk/lib/constructs/s3';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import {
	Effect,
	ManagedPolicy,
	PolicyStatement,
	ServicePrincipal,
} from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export interface CloudwatchLogsManagementProps
	extends Omit<GuStackProps, 'stage' | 'env'> {
	retentionInDays?: number;
	logShippingPrefixes?: string[];
}

export class CloudwatchLogsManagement extends GuStack {
	constructor(scope: App, props: CloudwatchLogsManagementProps) {
		const {
			stack,
			retentionInDays = 7,
			logShippingPrefixes = ['/aws/lambda'],
		} = props;

		// The ID will become `CloudwatchLogsManagement-<STACK>`
		const id = `${CloudwatchLogsManagement.prototype.constructor.name}-${stack}`;

		super(scope, id, {
			...props,
			stack,
			stage: 'PROD',
			env: {
				region: 'eu-west-1',
			},
		});

		const { region, account } = this;

		const kinesisStreamArn: string = new GuStringParameter(
			this,
			'KinesisStreamArn',
			{
				fromSSM: true,
				default: '/account/services/logging.stream',
				description: 'The ARN (not name) of the kinesis stream to ship logs to',
			},
		).valueAsString;

		const structuredFieldsBucket = new GuS3Bucket(
			this,
			'StructuredFieldsBucket',
			{ app: 'cloudwatch-logs-management' },
		);

		this.overrideLogicalId(structuredFieldsBucket, {
			logicalId: 'StructuredFieldsBucket',
			reason: 'Migrating from YAML',
		});

		const setRetentionLambda = new GuScheduledLambda(this, 'set-retention', {
			app: 'set-retention',
			runtime: Runtime.NODEJS_16_X,
			fileName: 'cloudwatch-logs-management.zip',
			handler: 'app.setRetention',
			rules: [{ schedule: Schedule.rate(Duration.hours(1)) }],
			monitoringConfiguration: { noMonitoring: true },
			environment: {
				RETENTION_IN_DAYS: retentionInDays.toString(),
			},
			timeout: Duration.minutes(1),
		});

		this.overrideLogicalId(setRetentionLambda, {
			logicalId: 'SetRetentionFunc',
			reason: 'Migrating from YAML',
		});

		const setRetentionPolicy = new ManagedPolicy(this, 'SetRetentionPolicy', {
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					actions: ['logs:DescribeLogGroups', 'logs:PutRetentionPolicy'],
					resources: [`arn:aws:logs:${region}:${account}:log-group:*`],
				}),
			],
		});
		setRetentionLambda.role?.addManagedPolicy(setRetentionPolicy);

		const shipLogEntriesLambda = new GuLambdaFunction(
			this,
			'ship-log-entries',
			{
				app: 'ship-log-entries',
				runtime: Runtime.NODEJS_16_X,
				fileName: 'cloudwatch-logs-management.zip',
				handler: 'shipLogEntries.shipLogEntries',
				timeout: Duration.seconds(5),
				environment: {
					LOG_KINESIS_STREAM: kinesisStreamArn,
					STRUCTURED_DATA_BUCKET: structuredFieldsBucket.bucketName,
				},
			},
		);

		this.overrideLogicalId(shipLogEntriesLambda, {
			logicalId: 'ShipLogEntriesFunc',
			reason: 'Migrating from YAML',
		});

		shipLogEntriesLambda.addPermission('ShipLogEntriesPermission', {
			principal: new ServicePrincipal(`logs.${region}.amazonaws.com`),
			sourceAccount: this.account,
		});

		const shipLogEntriesPolicies = [
			new ManagedPolicy(this, 'ShipLogEntriesPolicy', {
				statements: [
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['kinesis:PutRecords'],
						resources: [kinesisStreamArn],
					}),
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['s3:GetObject'],
						resources: [`${structuredFieldsBucket.bucketArn}/*`],
					}),
				],
			}),
			new ManagedPolicy(this, 'DisableCloudWatchLoggingPolicy', {
				statements: [
					new PolicyStatement({
						effect: Effect.DENY,
						actions: [
							'logs:CreateLogGroup',
							'logs:CreateLogStream',
							'logs:PutLogEvents',
						],
						resources: [`arn:aws:logs:*:*:*`],
					}),
				],
			}),
		];

		shipLogEntriesPolicies.forEach((policy) =>
			shipLogEntriesLambda.role?.addManagedPolicy(policy),
		);

		const setLogShippingLambda = new GuScheduledLambda(
			this,
			'set-log-shipping',
			{
				app: 'set-log-shipping',
				runtime: Runtime.NODEJS_16_X,
				fileName: 'cloudwatch-logs-management.zip',
				handler: 'app.setLogShipping',
				rules: [{ schedule: Schedule.rate(Duration.minutes(10)) }],
				monitoringConfiguration: { noMonitoring: true },
				environment: {
					LOG_SHIPPING_LAMBDA_ARN: shipLogEntriesLambda.functionArn,
					LOG_NAME_PREFIXES: logShippingPrefixes.join(','),
					STRUCTURED_DATA_BUCKET: structuredFieldsBucket.bucketName,
				},
				timeout: Duration.minutes(1),
			},
		);

		this.overrideLogicalId(setLogShippingLambda, {
			logicalId: 'SetLogShippingFunc',
			reason: 'Migrating from YAML',
		});

		const setLogShippingPolicy = new ManagedPolicy(
			this,
			'SetLogShippingPolicy',
			{
				statements: [
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: [
							'logs:DescribeLogGroups',
							'logs:DescribeSubscriptionFilters',
							'logs:PutSubscriptionFilter',
							'logs:DeleteSubscriptionFilter',
						],
						resources: [`arn:aws:logs:${region}:${account}:log-group:*`],
					}),
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['lambda:ListFunctions', 'lambda:ListTags'],
						resources: ['*'],
					}),
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['ecs:ListTaskDefinitions', 'ecs:DescribeTaskDefinition'],
						resources: ['*'],
					}),
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['s3:PutObject'],
						resources: [`${structuredFieldsBucket.bucketArn}/*`],
					}),
				],
			},
		);
		setLogShippingLambda.role?.addManagedPolicy(setLogShippingPolicy);
	}
}
