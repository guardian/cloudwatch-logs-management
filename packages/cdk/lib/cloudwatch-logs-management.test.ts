import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { stacks } from '../bin/cdk';
import { CloudwatchLogsManagement } from './cloudwatch-logs-management';
import { RiffRaffYamlFile } from './riff-raff-yaml';

describe('The CloudwatchLogsManagement stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new CloudwatchLogsManagement(app, { stack: 'deploy' });
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});

describe('The riff-raff.yaml file', () => {
	it('matches the snapshot', () => {
		const app = new App({ outdir: '/tmp/cdk.out' });
		stacks.forEach((stack) => new CloudwatchLogsManagement(app, stack));

		const riffRaff = new RiffRaffYamlFile(app);
		expect(riffRaff.toYAML()).toMatchSnapshot();
	});
});
