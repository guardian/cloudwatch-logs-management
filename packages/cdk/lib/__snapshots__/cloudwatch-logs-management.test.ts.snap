// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CloudwatchLogsManagement stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuS3Bucket",
      "GuStringParameter",
      "GuLambdaFunction",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "KinesisStreamArn": {
      "Default": "/account/services/logging.stream",
      "Description": "The ARN (not name) of the kinesis stream to ship logs to",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DisableCloudWatchLoggingPolicy566D41D3": {
      "Properties": {
        "Description": "",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Deny",
              "Resource": "arn:aws:logs:*:*:*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "SetLogShippingFunc": {
      "DependsOn": [
        "setlogshippingServiceRoleDefaultPolicy21BAA7D8",
        "setlogshippingServiceRoleFD6872A4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/PROD/set-log-shipping/set-log-shipping.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "set-log-shipping",
            "LOG_KINESIS_STREAM": {
              "Ref": "KinesisStreamArn",
            },
            "LOG_NAME_PREFIXES": "/aws/lambda",
            "LOG_SHIPPING_LAMBDA_ARN": {
              "Fn::GetAtt": [
                "ShipLogEntriesFunc",
                "Arn",
              ],
            },
            "STACK": "deploy",
            "STAGE": "PROD",
            "STRUCTURED_DATA_BUCKET": {
              "Ref": "StructuredFieldsBucket",
            },
          },
        },
        "Handler": "handlers.setLogShipping",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "setlogshippingServiceRoleFD6872A4",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "set-log-shipping",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SetLogShippingPolicy0F235CDA": {
      "Properties": {
        "Description": "",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:DescribeLogGroups",
                "logs:DescribeSubscriptionFilters",
                "logs:PutSubscriptionFilter",
                "logs:DeleteSubscriptionFilter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:logs:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "lambda:ListFunctions",
                "lambda:ListTags",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "ecs:ListTaskDefinitions",
                "ecs:DescribeTaskDefinition",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "StructuredFieldsBucket",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "SetRetentionFunc": {
      "DependsOn": [
        "setretentionServiceRoleDefaultPolicyCEB3697A",
        "setretentionServiceRoleF411F948",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/PROD/set-retention/set-retention.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "set-retention",
            "RETENTION_IN_DAYS": "7",
            "STACK": "deploy",
            "STAGE": "PROD",
          },
        },
        "Handler": "handlers.setRetention",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "setretentionServiceRoleF411F948",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "set-retention",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SetRetentionPolicy7B345D11": {
      "Properties": {
        "Description": "",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:DescribeLogGroups",
                "logs:PutRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:logs:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "ShipLogEntriesFunc": {
      "DependsOn": [
        "shiplogentriesServiceRoleDefaultPolicy372B39B1",
        "shiplogentriesServiceRoleB4E65CC2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/PROD/ship-log-entries/ship-log-entries.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "ship-log-entries",
            "LOG_KINESIS_STREAM": {
              "Ref": "KinesisStreamArn",
            },
            "STACK": "deploy",
            "STAGE": "PROD",
            "STRUCTURED_DATA_BUCKET": {
              "Ref": "StructuredFieldsBucket",
            },
          },
        },
        "Handler": "handlers.shipLogEntries",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "shiplogentriesServiceRoleB4E65CC2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "ship-log-entries",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ShipLogEntriesPolicyCE9979ED": {
      "Properties": {
        "Description": "",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "kinesis:PutRecords",
              "Effect": "Allow",
              "Resource": {
                "Ref": "KinesisStreamArn",
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "StructuredFieldsBucket",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "StructuredFieldsBucket": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "cloudwatch-logs-management",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "setlogshippingServiceRoleDefaultPolicy21BAA7D8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/PROD/set-log-shipping/set-log-shipping.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/set-log-shipping",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/set-log-shipping/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "setlogshippingServiceRoleDefaultPolicy21BAA7D8",
        "Roles": [
          {
            "Ref": "setlogshippingServiceRoleFD6872A4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "setlogshippingServiceRoleFD6872A4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Ref": "SetLogShippingPolicy0F235CDA",
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "set-log-shipping",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "setlogshippingsetlogshippingrate10minutes07B669287": {
      "Properties": {
        "ScheduleExpression": "rate(10 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "SetLogShippingFunc",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "setlogshippingsetlogshippingrate10minutes0AllowEventRuleCloudwatchLogsManagementdeploysetlogshipping1BFA79438197006B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SetLogShippingFunc",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "setlogshippingsetlogshippingrate10minutes07B669287",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "setretentionServiceRoleDefaultPolicyCEB3697A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/PROD/set-retention/set-retention.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/set-retention",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/set-retention/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "setretentionServiceRoleDefaultPolicyCEB3697A",
        "Roles": [
          {
            "Ref": "setretentionServiceRoleF411F948",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "setretentionServiceRoleF411F948": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Ref": "SetRetentionPolicy7B345D11",
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "set-retention",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "setretentionsetretentionrate1hour0AllowEventRuleCloudwatchLogsManagementdeploysetretentionD9A0E7537297E68D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SetRetentionFunc",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "setretentionsetretentionrate1hour0DA4C9A13",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "setretentionsetretentionrate1hour0DA4C9A13": {
      "Properties": {
        "ScheduleExpression": "rate(1 hour)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "SetRetentionFunc",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "shiplogentriesServiceRoleB4E65CC2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Ref": "ShipLogEntriesPolicyCE9979ED",
          },
          {
            "Ref": "DisableCloudWatchLoggingPolicy566D41D3",
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "ship-log-entries",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cloudwatch-logs-management",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "shiplogentriesServiceRoleDefaultPolicy372B39B1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/PROD/ship-log-entries/ship-log-entries.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/ship-log-entries",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/deploy/ship-log-entries/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "shiplogentriesServiceRoleDefaultPolicy372B39B1",
        "Roles": [
          {
            "Ref": "shiplogentriesServiceRoleB4E65CC2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "shiplogentriesShipLogEntriesPermission390E0555": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ShipLogEntriesFunc",
            "Arn",
          ],
        },
        "Principal": "logs.eu-west-1.amazonaws.com",
        "SourceAccount": {
          "Ref": "AWS::AccountId",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
